drivers/modem/modem_socket.c

```diff
@@ -275,29 +277,51 @@ int modem_socket_poll(struct modem_socket_config *cfg,
 			 * we consider the socket to always be writeable.
 			 */
 			if (fds[i].events & ZSOCK_POLLOUT) {
-				fds[i].revents |= ZSOCK_POLLOUT;
 				found_count++;
+				break;
 			} else if (fds[i].events & ZSOCK_POLLIN) {
 				sock->is_polled = true;
+
+				/*
+				 * Handle check done after data reception on
+				 * the socket. In this case that was received
+				 * but as the socket wasn't polled, no sem_poll
+				 * semaphore was given at that time. Therefore
+				 * if there is a polled socket with data,
+				 * increment found_count to escape the
+				 * k_sem_take().
+				 */
+				if (sock->packet_sizes[0] > 0U) {
+					found_count++;
+					break;
+				}
 			}
 		}
 	}
 
-	/* exit early if we've found rdy sockets */
-	if (found_count) {
-		errno = 0;
-		return found_count;
+	/* Avoid waiting on semaphore if we have already found an event */
+	ret = 0;
+	if (!found_count) {
+		ret = k_sem_take(&cfg->sem_poll, K_MSEC(msecs));
 	}
+	/* Reset counter as we reiterate on all polled sockets */
+	found_count = 0;
 
-	ret = k_sem_take(&cfg->sem_poll, K_MSEC(msecs));
 	for (i = 0; i < nfds; i++) {
 		sock = modem_socket_from_fd(cfg, fds[i].fd);
 		if (!sock) {
 			continue;
 		}
 
-		if ((fds[i].events & ZSOCK_POLLIN) &&
-		    (sock->packet_sizes[0] > 0U)) {
+		/*
+		 * Handle user check for ZSOCK_POLLOUT events:
+		 * we consider the socket to always be writeable.
```

Comment:
`s/writeable/writable/`

Reply:
You've missed this one.

Reply:
Exact, I fixed the wrong one. Now both should be ok!